### ISSUE TITLE
Sik-Ho Tang | Review: Layer Normalization (LN).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#108]
Sik-Ho Tang. [Review: Layer Normalization (LN)](https://sh-tsang.medium.com/review-layer-normalization-ln-6c2ae88bae47).

**Comment by NorbertZheng** [comment]
## Overview
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/4536896e-bf3d-4f38-b31f-18c7bf566a2c)
Layer Normalization (Image from [Group Normalization](https://sh-tsang.medium.com/review-group-norm-gn-group-normalization-image-classification-5f7fe0f58eb6)).

Layer Normalization. LN, by University of Toronto, and Google Inc. 2016 arXiv, Over 4000 Citations. Image Classification, Batch Normalization, Layer Normalization.

 - Batch Normalization ([BN](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651)) is dependent on the mini-batch size.
 - Layer Normalization (LN) is proposed by computing the mean and variance used for normalization from all of the summed inputs to the neurons in a layer on a single training case.
 - This is a tech report from Prof. Geoffrey E. Hinton Group.

**Comment by NorbertZheng** [comment]
## Batch Normalization ([BN](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651))
### Conventional Neural Network Without [BN](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651)
A feed-forward neural network is a non-linear mapping from a input pattern $x$ to an output vector $y$.

The summed inputs are computed through a linear projection with the weight matrix $W_{l}$ and the bottom-up inputs $h_{l}$ given as follows:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/27f86e54-f2bf-441f-9384-912f9a292e22)

where $b_{i}^{l}$ is bias.

**Comment by NorbertZheng** [comment]
### Conventional Neural Network With BN
[BN](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) was proposed in [BN-Inception / Inception-v2](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) to reduce undesirable “covariate shift”. The method normalizes the summed inputs to each hidden unit over the training cases.

Specifically, for the $i$-th summed input in the $l$-th layer, the batch normalization method rescales the summed inputs according to their variances under the distribution of the data:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/098b56f2-b8f4-45d5-8394-89e36712bc49)
where $\bar{a}_{i}^{l}$ is normalized summed inputs to the $i$-th hidden unit in the $l$-th layer and $g_{i}$ is a gain parameter scaling the normalized activation before the non-linear activation function.

 - **Note the expectation is under the whole training data distribution.**

$\mu$ and $\sigma$ are estimated using the empirical samples from the current mini-batch.

This puts constraints on the size of a mini-batch, and it is hard to apply to recurrent neural network (RNN).

**Comment by NorbertZheng** [comment]
## Layer Normalization (LN)
### LN
In LN, the “covariate shift” problem can be reduced by fixing the mean and the variance of the summed inputs **within each layer**.

 - The LN statistics is computed over all the hidden units in the same layer as follows:

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/8166900b-18a9-466b-af08-14700a90751a)
where $H$ denotes the number of hidden units in a layer.

 - Unlike [BN](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651), **LN does not impose any constraint on the size of a mini-batch** and it can be used in the pure online regime with batch size 1.

**Comment by NorbertZheng** [comment]
### Layer Normalized RNN
In a standard RNN, the summed inputs in the recurrent layer are computed from the current input $x_{t}$ and previous vector of hidden states $h_{t-1}$:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/374262d1-bc72-4a46-8e1d-ac4d22e6ce3a)

In a standard RNN, there is a tendency for the average magnitude of the summed inputs to the recurrent units to
 - **either grow or shrink at every time-step, leading to exploding or vanishing gradients**.

The layer normalized recurrent layer re-centers and re-scales its activations:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/2d100616-731a-4d99-a49e-5dbb3f144328)

where $W_{hh}$ is the recurrent hidden to hidden weights and $W_{xh}$ are the bottom up input to hidden weights. $\odot$ is the element-wise multiplication between two vectors. $b$ and $g$ are defined as the bias and gain parameters of the same dimension as $h_{t}$.

 - **LN in RNN results in much more stable hidden-to-hidden dynamics.**

**Comment by NorbertZheng** [comment]
## Experimental Results
### [Skip-Thought](https://sh-tsang.medium.com/review-skip-thought-vectors-a18565e7255a) Experiments
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/3835923a-75e0-462e-80b9-d1cff78cbfa9)
Performance of [Skip-Thought](https://sh-tsang.medium.com/review-skip-thought-vectors-a18565e7255a) with and without LN on downstream tasks as a function of training iterations.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/2e7f15fa-3f55-44f5-9c4a-451392a4a933)
[Skip-Thought](https://sh-tsang.medium.com/review-skip-thought-vectors-a18565e7255a) Results.

 - **Applying LN results both in speedup over the baseline as well as better final results after 1M iterations.**

**Comment by NorbertZheng** [comment]
### Permutation Invariant MNIST
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/52eb3fe6-420d-470f-9af1-c4bc8198beec)
Permutation invariant MNIST 784–1000–1000–10 model negative log likelihood and test error with layer normalization and batch normalization.

LN is only applied to the fully-connected hidden layers excluding the last softmax layer.

 - **LN is robust to the batch-sizes and exhibits a faster training convergence comparing to [BN](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) that is applied to all layers.**

**Comment by NorbertZheng** [comment]
In this paper, CNN is tried but the performance is not as good as fully connected network, and authors said **further research is needed to make LN work well in ConvNets**.

Yet, later research works already show that LN performs [BN](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) when batch size is small, or batches need to be distributed to multiple GPUs.

**Comment by NorbertZheng** [comment]
Layer Norm is used in many NLP models such as [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) and [Transformer-XL](https://sh-tsang.medium.com/review-transformer-xl-attentive-language-models-beyond-a-fixed-length-context-3168226cb5f).

**Comment by NorbertZheng** [comment]
## Reference
 - [2016 arXiv] [Layer Norm, LN] [Layer Normalization](https://arxiv.org/abs/1607.06450).
